### PR TITLE
Added support for title in get-Credential, and message set to optional.

### PR DESCRIPTION
**Pull request by McAndersDK**
Added possibility to add title to the credential prompt window.
Added possibility to only supply username without giving a message.

### CODE REVIEW

**Comment by lzybkr** [comment]
Your code does not compile.  Please fix before we review this PR.

**Comment by McAndersDK** [comment]
it build on my machine? 
I cant see in the build details what is wrong? @lzybkr

**Comment by vors** [comment]
@McAndersDK you can see logs from the build server by clicking on the build status in your PR (red cross)

For example, here
https://ci.appveyor.com/project/PowerShell/powershell/build/6.0.0-alpha.9-36#L276

The compilation error message

```
C:\projects\powershell\src\Microsoft.PowerShell.Security\security\CredentialCommands.cs(122,83): error CS1061: 'GetCredentialCommand' does not contain a definition for 'Username' and no extension method 'Username' accepting a 
first argument of type 'GetCre
dentialCommand' could be found (are you missing a using directive or an assembly reference?)
C:\projects\powershell\src\Microsoft.PowerShell.Security\security\CredentialCommands.cs(124,40): error CS1061: 'GetCredentialCommand' does not contain a definition for 'Username' and no extension method 'Username' accepting a 
first argument of type 'GetCre
dentialCommand' could be found (are you missing a using directive or an assembly reference?)
C:\projects\powershell\src\Microsoft.PowerShell.Security\security\CredentialCommands.cs(129,67): error CS0103: The name 'caption' does not exist in the current context
```

**Comment by McAndersDK** [comment]
Thanks @vors for some reason appveyor only showed the first 200 lines here.
I had to download the log to see the complete output.

**Comment by vors** [comment]
@McAndersDK it should not be the case in general, but could be a one-off problem with AppVeyor.

**Comment by lzybkr** [comment]
Thanks for fixing the build errors.

Before accepting this PR, we'll need some tests.  I was going to accept this PR without tests because `Get-Credential` is a bit harder to test than some cmdlets, but...

@JamesWTruher has ported our `Get-Credential` to Pester, hopefully we'll see a PR from him today.

After that PR is accepted, I'll ask you to add some tests for your new functionality.

File: src/Microsoft.PowerShell.Security/security/CredentialCommands.cs [hunk 1]
```diff
@@ -86,14 +112,22 @@ public GetCredentialCommand() : base()
...
-            if (!string.IsNullOrEmpty(this.Message))
+           string caption;
+            if(!string.IsNullOrEmpty(this.Title)) {
+                caption = this.Title;
+            }
+            else {
+                caption = UtilsStrings.PromptForCredential_DefaultCaption;
+            }
+            
+            if (!string.IsNullOrEmpty(this.Message) || !string.IsNullOrEmpty(this.UserName))
```
Comment (daxian-dbw) [inline src/Microsoft.PowerShell.Security/security/CredentialCommands.cs]: if only `-Title` is specified, the code path seems cannot get to `this.Host.UI.PromptForCredential`

File: src/Microsoft.PowerShell.Security/security/CredentialCommands.cs [hunk 2]
```diff
@@ -86,14 +112,22 @@ public GetCredentialCommand() : base()
         /// </summary>
         protected override void BeginProcessing()
         {
-            if (!string.IsNullOrEmpty(this.Message))
+           string caption;
```
Comment (daxian-dbw) [inline src/Microsoft.PowerShell.Security/security/CredentialCommands.cs]: please fix the indention before `string caption`

File: src/Microsoft.PowerShell.Security/security/CredentialCommands.cs [hunk 3]
```diff
@@ -16,7 +16,7 @@ namespace Microsoft.PowerShell.Commands
     /// operations involving security.
     /// </summary>
     [Cmdlet(VerbsCommon.Get, "Credential", DefaultParameterSetName = GetCredentialCommand.credentialSet, HelpUri = "http://go.microsoft.com/fwlink/?LinkID=113311")]
-    [OutputType(typeof(PSCredential), ParameterSetName = new string[] { GetCredentialCommand.credentialSet, GetCredentialCommand.messageSet })]
+    [OutputType(typeof(PSCredential), ParameterSetName = new string[] { GetCredentialCommand.credentialSet, GetCredentialCommand.messageSet, GetCredentialCommand.titleSet, GetCredentialCommand.usernameSet })]
```
Comment (daxian-dbw) [inline src/Microsoft.PowerShell.Security/security/CredentialCommands.cs]: It feels not necessary to create two more parameter sets.
How about making `Message` not mandatory? It won't be a breaking change to existing scripts.
Then you can add `Title` to the existing `MessageSet` and the `ValidateNotNullOrEmpty` attribute can be added to `Message` and `Title`.

File: src/Microsoft.PowerShell.Security/security/CredentialCommands.cs [hunk 4]
```diff
@@ -86,14 +112,22 @@ public GetCredentialCommand() : base()
...
         protected override void BeginProcessing()
         {
-            if (!string.IsNullOrEmpty(this.Message))
+           string caption;
+            if(!string.IsNullOrEmpty(this.Title)) {
+                caption = this.Title;
+            }
+            else {
+                caption = UtilsStrings.PromptForCredential_DefaultCaption;
+            }
```
Comment (daxian-dbw) [inline src/Microsoft.PowerShell.Security/security/CredentialCommands.cs]: If you add `ValidateNotNullOrEmpty` attribute to `Title`, then this can be turned to
```
string caption = this.Title ?? UtilsStrings.PromptForCredential_DefaultCaption; 
```

**Comment by McAndersDK** [comment]
@daxian-dbw Thanks for input, I will make sure this will go into the PR with the test for the changes.

**Comment by jpsnover** [comment]
We should have done this a long time ago. Thanks!

File: src/Microsoft.PowerShell.Security/security/CredentialCommands.cs [hunk 5]
```diff
@@ -86,14 +112,22 @@ public GetCredentialCommand() : base()
         /// </summary>
         protected override void BeginProcessing()
         {
-            if (!string.IsNullOrEmpty(this.Message))
+           string caption;
+            if(!string.IsNullOrEmpty(this.Title)) {
```
Comment (rkeithhill) [inline src/Microsoft.PowerShell.Security/security/CredentialCommands.cs]: Looks like this code based uses a space between the `if` keyword and the opening paren. To be consistent with the rest of the code, you should add the space.

**Comment by lzybkr** [comment]
Can you add a test to test\powershell\Modules\Microsoft.PowerShell.Security\GetCredential.Tests.ps1 and fix the formatting as @rkeithhill suggested (though we'd fix that with a tool later anyway).

**Comment by McAndersDK** [comment]
yeah I will. Should I create a new PR with both file in it or just create a new one with the test separate?

**Comment by lzybkr** [comment]
This PR is fine - just push another commit to your branch.

File: src/Microsoft.PowerShell.Security/security/CredentialCommands.cs [hunk 6]
```diff
@@ -54,24 +54,38 @@ public PSCredential Credential
...
             set { _message = value; }
         }
-        private string _message = null;
+        private string _message;
 
         /// <summary>
         /// Gets and sets the user supplied username to be used while creating the PSCredential.
         /// </summary>
         [Parameter(Position = 0, Mandatory = false, ParameterSetName = messageSet)]
+        [ValidateNotNullOrEmpty]
```
Comment (lzybkr) [inline src/Microsoft.PowerShell.Security/security/CredentialCommands.cs]: This is a breaking change.  A script might use something like:
```PowerShell
Get-Credential -User $sser
````
Note the typo, `$user` was intended.  This script would still run and prompt, just without the user filled in already.

With your change, the script would fail with an unexpected error.

File: src/Microsoft.PowerShell.Security/security/CredentialCommands.cs [hunk 7]
```diff
@@ -86,20 +100,18 @@ public GetCredentialCommand() : base()
         /// </summary>
         protected override void BeginProcessing()
         {
-            if (!string.IsNullOrEmpty(this.Message))
-            {
-                string caption = UtilsStrings.PromptForCredential_DefaultCaption;
-                string message = this.Message;
+            string caption = this.Title ?? UtilsStrings.PromptForCredential_DefaultCaption;
+            string message = this.Message ?? "Enter your credentials.";
```
Comment (lzybkr) [inline src/Microsoft.PowerShell.Security/security/CredentialCommands.cs]: "Enter your credentials." cannot be localized - it needs to be in a resource.

Comment (lzybkr) [inline src/Microsoft.PowerShell.Security/security/CredentialCommands.cs]: Maybe it's better to leave Message mandatory?

**Comment by McAndersDK** [comment]
@JamesWTruher How do I access the prompt stream?

File: src/Microsoft.PowerShell.Security/security/CredentialCommands.cs (hunk 7 above)
Comment (McAndersDK) [inline src/Microsoft.PowerShell.Security/security/CredentialCommands.cs]: Ah yeah true. so just:
string message = this.Message ?? null;
?
I dont like that the message is mandatory, as if one just want to specifi username or the title.
where is no reason for message to be mandatory?

File: src/Microsoft.PowerShell.Security/security/CredentialCommands.cs [hunk 8]
```diff
@@ -86,20 +100,18 @@ public GetCredentialCommand() : base()
...
+            string username = this.UserName ?? null;
```
Comment (daxian-dbw) [inline src/Microsoft.PowerShell.Security/security/CredentialCommands.cs]: > string username = this.UserName ?? null; 

This is not necessary. Plus, it seems `username` is never used anywhere.

File: src/Microsoft.PowerShell.Security/security/CredentialCommands.cs (hunk 6 above)
Comment (McAndersDK) [inline src/Microsoft.PowerShell.Security/security/CredentialCommands.cs]: that is true, but that go with title and message also.
Ok so lets change them alle to
string title
if (this.MyInvocation.BoundParameters.ContainsKey("UserName")) 
{
username = this.UserName
}
else
{
title = Title
}
?

File: src/Microsoft.PowerShell.Security/security/CredentialCommands.cs (hunk 7 above)
Comment (daxian-dbw) [inline src/Microsoft.PowerShell.Security/security/CredentialCommands.cs]: you can add the default message to `Microsoft.PowerShell.Security\resources\UtilsStrings.resx`. You can find the `PromptForCredential_DefaultCaption` defined there.

File: src/Microsoft.PowerShell.Security/security/CredentialCommands.cs (hunk 8 above)
Comment (McAndersDK) [inline src/Microsoft.PowerShell.Security/security/CredentialCommands.cs]: That should have been used on line 96 not _userName

Comment (daxian-dbw) [inline src/Microsoft.PowerShell.Security/security/CredentialCommands.cs]: If you are not going to use a different value when `this.UserName` is null, then `this.UserName ?? null;` is not necessary anymore. Line 96 can continue to use _userName.

File: src/Microsoft.PowerShell.Security/security/CredentialCommands.cs (hunk 6 above)
Comment (daxian-dbw) [inline src/Microsoft.PowerShell.Security/security/CredentialCommands.cs]: for `Message` parameter, it was mandatory, and that implies `NotNullOrEmpty`, so adding `[ValidateNotNullOrEmpty]` to `Message` is fine. As for `Title`, it's new, so adding `[ValidateNotNullOrEmpty]` is also fine.
You just need to remove the `[ValidateNotNullOrEmpty]` attribute from `UserName`.

**Comment by McAndersDK** [comment]
@jpsnover I agree 👍 
I accept any PowerShell swags you can collect 😄

**Comment by daxian-dbw** [comment]
:shipit:

**Comment by LeeHolmes** [comment]
One danger this opens Get-Credential up to is a pretty convincing phishing vehicle - for example, if the title is "Microsoft Outlook", you might not realize this is actually from a PowerShell script.

When an alternate title is supplied, I suggest the following text - similar to what we do when there is a credential request coming over a remoting channel:

Warning: A PowerShell script is requesting your credentials. Enter your credentials only if you trust the script that is requesting them.

**Comment by McAndersDK** [comment]
@LeeHolmes  that is true, but this change will just make it a tiny bit easyer to do.
nothing stop them from using
``$Host.UI.PromptForCredential('Microsoft outlook','','','')

but to be trully good, one would use YAML to create a 100% look alike to the outlook credential prompt.

**Comment by lzybkr** [comment]
@McAndersDK - in the test host, it looks like you can use `$th.UI.Streams.Prompt`, see https://github.com/PowerShell/PowerShell/blob/master/test/powershell/Common/TestHostCS.psm1#L129

**Comment by LeeHolmes** [comment]
@McAndersDK - That's a good point, moving the prompt warning to the default host is probably the best place to put it.

**Comment by jpsnover** [comment]
I struggled with this Q and acknowledging that a bad guy can work around things - I wonder our cmdlet should have a prefix for whatever title string a user supplies.  

I don't know what the right prefix would be but for this example, image that we choose "XXXX"
e.g. Get-Credential -Title "ENTER OUTLOOK CREDS"
would have a title :
"XXXX: ENTER OUTLOOK CREDS"

thoughts?

**Comment by McAndersDK** [comment]
For my use case, prepending a string Will also Work. 

The string could also just be the default caption, and the 'title' parameter only get added at the end of the default.

**Comment by McAndersDK** [comment]
I also agree with @leeholmes that the host should be the one controlling the title, so one cant bypass the this. 

A solution could also be to make the credentials Prompt stand out, so every one would know it came from the powershell host.

**Comment by lzybkr** [comment]
I'm fine with opening another issue to make the host ensure the prompt is unambiguously coming from PowerShell.

@McAndersDK - I think this is ready merge unless you're adding a test to verify the title is passed correctly to the host.

Such a test would be nice but not required as it's still insufficient in validating the title gets displayed, the only way to validate that would require an `expect` like test.

**Comment by McAndersDK** [comment]
Yeah i need to add some more tests with the streams. 

Please give me some days to add them in.

**Comment by lzybkr** [comment]
Awesome - thanks for being responsive to our feedback.

**Comment by McAndersDK** [comment]
Travis just timed out?
How do we rerun travis? except to do a push?

File: test/powershell/Modules/Microsoft.PowerShell.Security/GetCredential.Tests.ps1 [hunk 9]
```diff
@@ -21,11 +21,41 @@ Describe "Get-Credential Test" -tag "CI" {
...
-        $cred = $ps.AddScript("Get-Credential -User Joe -Message Foo").Invoke() | Select-Object -First 1
+    It "Get-Credential with message, produces a credential object" {
+        $cred = $ps.AddScript("Get-Credential -UserName Joe -Message Foo").Invoke() | Select-Object -First 1
         $cred.gettype().FullName | Should Be "System.Management.Automation.PSCredential"
         $netcred = $cred.GetNetworkCredential()
         $netcred.UserName | Should be "Joe"
         $netcred.Password | Should be "this is a test"
+        # Resx file only include english language, this should be changed to reflect Current Culture if title change by language.
+        $th.ui.Streams.Prompt[-1] | should be "Credential:Windows PowerShell credential request.:Foo" 
+    }
```
Comment (lzybkr) [inline test/powershell/Modules/Microsoft.PowerShell.Security/GetCredential.Tests.ps1]: We do run tests on localized builds of Windows, so this test would fail on those builds.

In some tests, if the message is important, we'll actually load the resource to get the text - that works on localized builds as well. We might not have any examples of that pattern in out pester tests though.

File: test/powershell/Modules/Microsoft.PowerShell.Security/GetCredential.Tests.ps1 [hunk 10]
```diff
@@ -21,11 +21,41 @@ Describe "Get-Credential Test" -tag "CI" {
...
+        $cred = $ps.AddScript("Get-Credential -UserName Joe -Message Foo -Title CustomTitle").Invoke() | Select-Object -First 1
+        $cred.gettype().FullName | Should Be "System.Management.Automation.PSCredential"
+        $netcred = $cred.GetNetworkCredential()
+        $netcred.UserName | Should be "Joe"
+        $netcred.Password | Should be "this is a test"
+        $th.ui.Streams.Prompt[-1] | should be "Credential:CustomTitle:Foo"
+    }
+    It "Get-Credential with empty username, should throw" {
+        $cred = $ps.AddScript("Get-Credential -UserName ''").Invoke() | should throw
     }
```
Comment (lzybkr) [inline test/powershell/Modules/Microsoft.PowerShell.Security/GetCredential.Tests.ps1]: A 'should throw' test always has pipes a scriptblock, e.g.

```powershell
{ $ps.AddScript("Get-Credential -UserName ''").Invoke() } | should throw
```

**Comment by McAndersDK** [comment]
@lzybkr That did not work. we need to test the invoke for credential object?, if none throw?
but why change this? it worked before? atleast the test did not fail, I have not tryed the other way around

**Comment by lzybkr** [comment]
The problem with not using a script block is that if it doesn't throw, the test passes but should fail. Adding the change I suggested exposed the fact that no exception was thrown, so failing the test is correct.

I think this test might not make sense.  Looking at the console host - we prompt for the username if it's null or empty. Other hosts might do that or throw, but we don't really control other hosts.

I'd say just remove this test.

**Review by lzybkr — APPROVED** [review]

**Comment by iSazonov** [comment]
Set "message" to optional is not best practice, from https://msdn.microsoft.com/en-us/library/dd878348%28v=vs.85%29.aspx :
•Each parameter set must have at least one unique parameter. If possible, make this parameter a **mandatory** parameter.

Set both "message" and "title" to optional open a way to "anonymous" request of credential. That is bad. This provokes writing bad code.

**Comment by lzybkr** [comment]
Before this change, you could call `Get-Credential` with no parameters, so I thought the change was necessary to keep that behavior. Maybe it wasn't.

The current experience (after this PR) with `Get-Credential` and no additional parameters is broken - it prompts twice (see #2306). So maybe we need to revert the change on the `Message` parameter.

**Comment by iSazonov** [comment]
> Before this change, you could call  Get-Credential  with no parameters
Get-Credential  with no parameters set message to default "Enter you credentials". Message is in another parameter set and that is another experience.
So I think we should follow the msdn recommendations above mentioned.
